### PR TITLE
feat(config): migrate to Talos v1.12 multi-document config format

### PR DESCRIPTION
**Pull request by lexfrei**
## Summary

Add support for the Talos v1.12 multi-document configuration format. When `TalosVersion` is set to v1.12 or later (via `Chart.yaml` or `--talos-version` CLI flag), templates generate separate YAML documents instead of the deprecated monolithic `machine.network` and `machine.registries` fields.

Closes #100

### Engine
- Pass `TalosVersion` through the Helm values map into the template rendering context (available as `.TalosVersion`), avoiding global mutable state

### Chart templates (cozystack + generic)
- Split `talos.config` into shared sub-templates to eliminate duplication: `talos.config.machine.common`, `talos.config.cluster`, `talos.config.network.legacy`, `talos.config.network.multidoc`
- Version dispatcher selects legacy or multi-doc format based on `semverCompare`
- New v1.12 document types generated: `HostnameConfig`, `ResolverConfig`, `LinkConfig`, `BondConfig`, `VLANConfig`, `RegistryMirrorConfig`, `Layer2VIPConfig`

### Bug fix
- Fix `nr_hugepages` rendering inside `with` block (was producing empty string for non-zero values due to `.Values` being inaccessible when dot is rebound)

### Backward compatibility
- Legacy format is fully preserved when `TalosVersion` is empty or < v1.12
- All legacy regression tests pass unchanged

### Tests
- `isTalosConfigPatch` and `extractExtraDocuments` tests for all 7 new document types
- Offline rendering tests for both charts x both roles x both formats
- Version edge cases: pre-release, two-component version strings
- Concurrent render test with race detector
- `nr_hugepages` rendering test for both legacy and multi-doc paths

## Summary by CodeRabbit

* **Documentation**
  * Clarified Talos config output differences: Talos < v1.12 emits a single YAML document; Talos ≥ v1.12 emits a multi-document stream. Notes added for conditional Layer2 VIP and registry-mirror emission.

* **Refactor**
  * Chart rendering now branches between legacy and multi-document Talos formats, centralizing shared machine/cluster content and adapting network output per Talos version.

* **Tests**
  * Expanded rendering tests, added Talos-version validation and concurrent/template-context checks.

### CODE REVIEW

File: charts/cozystack/templates/_helpers.tpl [hunk 1]
```diff
@@ -161,3 +122,183 @@ cluster:
...
+kind: ResolverConfig
+nameservers:
+{{- $resolvers := include "talm.discovered.default_resolvers" . }}
+{{- if $resolvers }}
+{{- range fromJsonArray $resolvers }}
+  - address: {{ . | quote }}
+{{- end }}
+{{- else }}
+  []
+{{- end }}
```
[the remark this answers is not in the document]
Comment (lexfrei) [reply charts/cozystack/templates/_helpers.tpl:148]: The current format is correct per the Go type definitions.

`ResolverConfigV1Alpha1` has `ResolverNameservers []NameserverConfig` with yaml tag `nameservers,omitempty`, where `NameserverConfig` has `Address Addr` with yaml tag `address`.

Source: [`resolver.go:63`](https://github.com/siderolabs/talos/blob/main/pkg/machinery/config/types/network/resolver.go) and [`resolver.go:81`](https://github.com/siderolabs/talos/blob/main/pkg/machinery/config/types/network/resolver.go)

The official test data (`testdata/resolverconfig.yaml`) confirms the format:
```yaml
nameservers:
  - address: 10.0.0.1
  - address: 2001:4860:4860::8888
```

File: charts/cozystack/templates/_helpers.tpl [hunk 2]
```diff
@@ -161,3 +122,183 @@ cluster:
...
+{{- if not $isVlan }}
+addresses:
+{{- range fromJsonArray (include "talm.discovered.default_addresses_by_gateway" .) }}
+  - address: {{ . }}
+{{- end }}
+routes:
+  - gateway: {{ include "talm.discovered.default_gateway" . }}
+{{- end }}
+{{- end }}
```
[the remark this answers is not in the document]
Comment (lexfrei) [reply charts/cozystack/templates/_helpers.tpl:197]: Schema claims are incorrect, but the lookup safety point is valid.

**`bondMode` is correct** — `BondConfigV1Alpha1` has `BondMode *nethelpers.BondMode` with yaml tag `bondMode,omitempty` ([bond.go:96](https://github.com/siderolabs/talos/blob/main/pkg/machinery/config/types/network/bond.go)). Test data confirms: `bondMode: 802.3ad`.

**`addresses` format is correct** — `AddressConfig` has `AddressAddress netip.Prefix` with yaml tag `address` ([link.go:122](https://github.com/siderolabs/talos/blob/main/pkg/machinery/config/types/network/link.go)). Format is `- address: IP/prefix`, not bare strings.

**`routes` without `destination` is valid** — `RouteDestination` has yaml tag `destination,omitempty` (NOT `network`). The field is optional: _"If not specified, a default route will be created for the address family of the gateway."_ ([link.go:133-135](https://github.com/siderolabs/talos/blob/main/pkg/machinery/config/types/network/link.go))

**Lookup safety** — `$isBondInterface` calls `talm.discovered.is_bond` which already performs `lookup "links" "" $interfaceName` and checks `$link.spec.kind == "bond"`. If it returned true, the resource is guaranteed to exist. Nevertheless, adding a guard is reasonable defensive coding. Will fix.

File: charts/cozystack/templates/_helpers.tpl [hunk 3]
```diff
@@ -161,3 +122,183 @@ cluster:
...
+kind: VLANConfig
+name: {{ $defaultLinkName }}
+vlanID: {{ include "talm.discovered.vlan_id" $defaultLinkName }}
+parent: {{ $interfaceName }}
+addresses:
+{{- range fromJsonArray (include "talm.discovered.default_addresses_by_gateway" .) }}
+  - address: {{ . }}
+{{- end }}
+routes:
+  - gateway: {{ include "talm.discovered.default_gateway" . }}
```
[the remark this answers is not in the document]
Comment (lexfrei) [reply charts/cozystack/templates/_helpers.tpl:211]: All three claims are incorrect.

**`vlanID` is correct** — `VLANConfigV1Alpha1` has `VLANIDConfig uint16` with yaml tag `vlanID,omitempty` ([vlan.go:69](https://github.com/siderolabs/talos/blob/main/pkg/machinery/config/types/network/vlan.go)). Test data confirms: `vlanID: 2`.

**`addresses` format is correct** — uses `AddressConfig` with `address` key, same as LinkConfig/BondConfig.

**`routes` without `destination` is valid** — `RouteDestination` is `omitempty`, not required.

File: charts/cozystack/templates/_helpers.tpl [hunk 4]
```diff
@@ -161,3 +122,183 @@ cluster:
...
+---
+apiVersion: v1alpha1
+kind: LinkConfig
+name: {{ $interfaceName }}
+addresses:
+{{- range fromJsonArray (include "talm.discovered.default_addresses_by_gateway" .) }}
+  - address: {{ . }}
+{{- end }}
+routes:
+  - gateway: {{ include "talm.discovered.default_gateway" . }}
```
[the remark this answers is not in the document]
Comment (lexfrei) [reply charts/cozystack/templates/_helpers.tpl:222]: Both claims are incorrect.

**`addresses`** uses `AddressConfig` with yaml tag `address` — format is `- address: IP/prefix`, not bare strings ([link.go:98,122](https://github.com/siderolabs/talos/blob/main/pkg/machinery/config/types/network/link.go)).

**`routes`** — `RouteDestination` has yaml tag `destination,omitempty`. The `destination` field is optional, not required. There is no `network` field in `RouteConfig`. A route with only `gateway` creates a default route.

File: charts/cozystack/templates/_helpers.tpl [hunk 5]
```diff
@@ -161,3 +122,183 @@ cluster:
...
+{{- end }}
+{{- if and .Values.floatingIP (eq .MachineType "controlplane") }}
+---
+apiVersion: v1alpha1
+kind: Layer2VIPConfig
+name: {{ .Values.floatingIP | quote }}
+link: {{ $interfaceName }}
```
[the remark this answers is not in the document]
Comment (lexfrei) [reply charts/cozystack/templates/_helpers.tpl]: Incorrect. The VIP IP address is stored in the `name` field.

`Layer2VIPConfigV1Alpha1` has `MetaName string` with yaml tag `name` ([layer2_vip.go:65](https://github.com/siderolabs/talos/blob/main/pkg/machinery/config/types/network/layer2_vip.go)). The `VIP()` method parses this field as an IP address ([layer2_vip.go:128-131](https://github.com/siderolabs/talos/blob/main/pkg/machinery/config/types/network/layer2_vip.go)). Validation at line 111 confirms `MetaName` must be a valid IP.

Test data confirms: `name: 1.2.3.4`. There is no `ip` field in this type.

File: charts/cozystack/templates/_helpers.tpl [hunk 6]
```diff
@@ -161,3 +122,183 @@ cluster:
...
+{{- define "talos.config.multidoc" }}
+{{- include "talos.config.machine.common" . }}
+
+{{- include "talos.config.cluster" . }}
+---
+apiVersion: v1alpha1
+kind: RegistryMirrorConfig
+name: docker.io
+endpoints:
+  - url: https://mirror.gcr.io
```
[the remark this answers is not in the document]
Comment (lexfrei) [reply charts/cozystack/templates/_helpers.tpl:308]: Incorrect. Endpoints are objects, not strings.

`RegistryMirrorConfigV1Alpha1` has `RegistryEndpoints []RegistryEndpoint` with yaml tag `endpoints,omitempty`. `RegistryEndpoint` has `EndpointURL meta.URL` with yaml tag `url` AND `EndpointOverridePath *bool` with yaml tag `overridePath,omitempty` ([registry_mirror.go:69,87,90](https://github.com/siderolabs/talos/blob/main/pkg/machinery/config/types/cri/registry_mirror.go)).

Endpoints must be objects because they support `overridePath`. Test data confirms:
```yaml
endpoints:
  - url: https://my-private-registry.local:5000
  - url: http://my-harbor/v2/registry-k8s.io/
    overridePath: true
```

File: charts/generic/templates/_helpers.tpl [hunk 7]
```diff
@@ -12,6 +21,138 @@ machine:
...
+kind: ResolverConfig
+nameservers:
+{{- $resolvers := include "talm.discovered.default_resolvers" . }}
+{{- if $resolvers }}
+{{- range fromJsonArray $resolvers }}
+  - address: {{ . | quote }}
+{{- end }}
+{{- else }}
+  []
+{{- end }}
```
[the remark this answers is not in the document]
Comment (lexfrei) [reply charts/generic/templates/_helpers.tpl:71]: Same as above — the format is correct. `NameserverConfig` has `Address Addr` with yaml tag `address`. See [resolver.go:63,81](https://github.com/siderolabs/talos/blob/main/pkg/machinery/config/types/network/resolver.go).

File: charts/generic/templates/_helpers.tpl [hunk 8]
```diff
@@ -12,6 +21,138 @@ machine:
...
+{{- if not $isVlan }}
+addresses:
+{{- range fromJsonArray (include "talm.discovered.default_addresses_by_gateway" .) }}
+  - address: {{ . }}
+{{- end }}
+routes:
+  - gateway: {{ include "talm.discovered.default_gateway" . }}
+{{- end }}
+{{- end }}
```
[the remark this answers is not in the document]
Comment (lexfrei) [reply charts/generic/templates/_helpers.tpl:120]: Schema claims are incorrect (see replies on the cozystack counterpart). The lookup safety point is valid though — adding a guard. Fixed.

File: charts/generic/templates/_helpers.tpl [hunk 9]
```diff
@@ -12,6 +21,138 @@ machine:
...
+kind: VLANConfig
+name: {{ $defaultLinkName }}
+vlanID: {{ include "talm.discovered.vlan_id" $defaultLinkName }}
+parent: {{ $interfaceName }}
+addresses:
+{{- range fromJsonArray (include "talm.discovered.default_addresses_by_gateway" .) }}
+  - address: {{ . }}
+{{- end }}
+routes:
+  - gateway: {{ include "talm.discovered.default_gateway" . }}
```
[the remark this answers is not in the document]
Comment (lexfrei) [reply charts/generic/templates/_helpers.tpl:134]: All claims incorrect — see replies on the cozystack counterpart. `vlanID` is correct yaml tag, `addresses` uses `AddressConfig` objects, `routes.destination` is optional.

**Review by myasnikovdaniil — COMMENTED** [review]
## Code Review

Nice work on the v1.12 multi-document config support — the version dispatching logic is correct, the `nr_hugepages` bugfix is solid, and the test matrix (2 charts × 2 roles × 2 formats) is thorough.

### Testing

Built the PR binary and tested against the dev10 cluster (3× controlplane, Talos v1.11.3):

| Test | Result |
|------|--------|
| Legacy rendering (v1.11) — PR vs main | Identical on all 3 nodes |
| Dry-run apply on all 3 nodes | Identical diffs between main and PR binary |
| Multi-doc offline (cozystack preset, CP + worker) | Correct — all expected document types emitted |
| `--talos-version=latest` on cozystack chart | Crashes with opaque error (see bug below) |

### Bug

**Malformed `TalosVersion` causes opaque template crash**

A non-semver `TalosVersion` (e.g. `latest`, `foobar`) passes the `and .TalosVersion` truthiness check but `semverCompare` returns an error, producing:

```
error calling semverCompare: invalid semantic version
```

The clean validation via `ParseContractFromVersion` in `engine.go` runs *after* template rendering, so it never gets a chance to produce a user-friendly error. Custom templates that don't use `semverCompare` reach the later validation and get a better message: `invalid talos-version: error parsing version "vlatest"`.

**Suggestion:** Validate `TalosVersion` before `eng.Render()` — e.g. call `semver.NewVersion(opts.TalosVersion)` early in `Render()` and return a clean error.

### Test coverage gaps

**BondConfig and VLANConfig paths are never exercised**

The offline `LookupFunc` always returns empty maps, so `is_bond`/`is_vlan` always evaluate to false. No test ever emits a `BondConfig` or `VLANConfig` document. The `hasLinkConfig || hasBondConfig` assertion always passes via `LinkConfig`. Consider adding tests with a custom `LookupFunc` returning mock bond/vlan resources.

**Layer2VIPConfig is never positively asserted**

The cozystack defaults include `floatingIP: 192.168.100.10`, so `Layer2VIPConfig` *is* present in the controlplane multi-doc output — but no test checks for it. All four `Layer2VIPConfig` references are `assertNotContains`. Adding `assertContains(t, output, "kind: Layer2VIPConfig")` to `TestMultiDocCozystack_ControlPlane` would close this gap.

### Nits

**README note could be more precise** — lists all 7 document types as always-present, but `BondConfig`/`VLANConfig`/`Layer2VIPConfig` are conditional on topology, and `RegistryMirrorConfig` is cozystack-only. A qualifier like "depending on network topology and chart preset" would help.

**Comment on `talos.config.machine.common` (cozystack) omits `nodeLabels`** — says "type, kubelet, sysctls, kernel, certSANs, files, install" but the template also renders `nodeLabels` for controlplane nodes.

**Review by myasnikovdaniil — COMMENTED** [review]
### Bug

**Malformed `TalosVersion` causes opaque template crash**

A non-semver `TalosVersion` (e.g. `latest`, `foobar`) passes the `and .TalosVersion` truthiness check but `semverCompare` returns an error, producing:

```
error calling semverCompare: invalid semantic version
```

The clean validation via `ParseContractFromVersion` in `engine.go` runs *after* template rendering, so it never gets a chance to produce a user-friendly error. Custom templates that don't use `semverCompare` reach the later validation and get a better message: `invalid talos-version: error parsing version "vlatest"`.

**Suggestion:** Validate `TalosVersion` before `eng.Render()` — e.g. call `semver.NewVersion(opts.TalosVersion)` early in `Render()` and return a clean error.

### Test coverage gaps

**BondConfig and VLANConfig paths are never exercised**

The offline `LookupFunc` always returns empty maps, so `is_bond`/`is_vlan` always evaluate to false. No test ever emits a `BondConfig` or `VLANConfig` document. The `hasLinkConfig || hasBondConfig` assertion always passes via `LinkConfig`. Consider adding tests with a custom `LookupFunc` returning mock bond/vlan resources.

**Layer2VIPConfig is never positively asserted**

The cozystack defaults include `floatingIP: 192.168.100.10`, so `Layer2VIPConfig` *is* present in the controlplane multi-doc output — but no test checks for it. All four `Layer2VIPConfig` references are `assertNotContains`. Adding `assertContains(t, output, "kind: Layer2VIPConfig")` to `TestMultiDocCozystack_ControlPlane` would close this gap.

### Nits

**README note could be more precise** — lists all 7 document types as always-present, but `BondConfig`/`VLANConfig`/`Layer2VIPConfig` are conditional on topology, and `RegistryMirrorConfig` is cozystack-only. A qualifier like "depending on network topology and chart preset" would help.

**Comment on `talos.config.machine.common` (cozystack) omits `nodeLabels`** — says "type, kubelet, sysctls, kernel, certSANs, files, install" but the template also renders `nodeLabels` for controlplane nodes.

**Comment by lexfrei** [comment]
**Addressed review feedback. Force-pushed on rebase onto main.**

### Changes since last review

**🐛 Bug — invalid `TalosVersion` crash**
`Render()` now calls `config.ParseContractFromVersion(opts.TalosVersion)` before template evaluation. Values like `latest` or `foobar` now fail with `invalid talos-version: error parsing version "vlatest"` instead of the opaque `semverCompare: invalid semantic version`.

**📊 Coverage — BondConfig / VLANConfig**
Added four new tests with mock `LookupFunc`:
- `TestMultiDoc{Cozystack,Generic}_BondTopology` — asserts `BondConfig` emission with slaves, `bondMode`, xmit hash/LACP, `addresses`, default route.
- `TestMultiDoc{Cozystack,Generic}_VlanOnBondTopology` — asserts combined `BondConfig` + `VLANConfig` emission with `vlanID`, `parent`, addresses.

**📊 Coverage — Layer2VIPConfig positive assert**
`TestMultiDocCozystack_ControlPlane` now contains `assertContains(output, "kind: Layer2VIPConfig")` and `assertContains(output, "192.168.100.10")`.

**🧹 Test isolation**
`TestLegacyCozystack_NrHugepages` and `TestMultiDocCozystack_NrHugepages` now restore `helmEngine.LookupFunc` via `t.Cleanup`.

**📝 README**
Clarified that only `HostnameConfig`, `ResolverConfig`, and one network interface document are always emitted; `Layer2VIPConfig` is controlplane-only when `floatingIP` is set; `RegistryMirrorConfig` is cozystack-only.

**📝 Comment fix**
`talos.config.machine.common` (cozystack) comment updated to include `nodeLabels`.

**🔒 Lookup safety (already in place)**
`{{- if $link }}` guard after `lookup "links" "" $interfaceName` in both `_helpers.tpl` — confirmed present on current HEAD.

### Gemini feedback

All schema claims from the bot-review (bare-string addresses, `mode`/`vlanId`/`ip`/`url`-list/string-nameserver) have been verified against the `siderolabs/talos` Go types and test data — the existing YAML shape is correct. References are in the inline replies on the original comments.

Tests: `go test -race ./...` green, `go vet` clean, `golangci-lint run` — 0 issues.

File: charts/cozystack/templates/_helpers.tpl [hunk 10]
```diff
@@ -161,3 +122,185 @@ cluster:
       {{- toYaml .Values.advertisedSubnets | nindent 6 }}
   {{- end }}
 {{- end }}
+
+{{- /* Shared network document generation for v1.12+ multi-doc format */ -}}
+{{- define "talos.config.network.multidoc" }}
+{{- /* Multi-doc format always reconstructs network config from discovery resources.
+       existing_interfaces_configuration is not used here because v1.12 nodes store
+       network config in separate documents (LinkConfig, BondConfig, etc.), not in
+       the legacy machine.network.interfaces field. */ -}}
```
[the remark this answers is not in the document]
Comment (lexfrei) [reply charts/cozystack/templates/_helpers.tpl:131]: Intentional. `talm.discovered.existing_interfaces_configuration` emits a legacy `machine.network.interfaces` YAML fragment, which has no safe 1:1 translation to the v1.12 multi-doc resources (LinkConfig / BondConfig / VLANConfig / Layer2VIPConfig). Rather than silently producing a partial or incorrect translation, the multidoc path rebuilds the required documents from the discovery resources directly; the comment at the top of `talos.config.network.multidoc` documents this. Users who rely on custom legacy interface state should keep `machine.talosVersion` on `v1.11.x` until they migrate their overrides to the new resources.

File: charts/cozystack/templates/_helpers.tpl [hunk 11]
```diff
@@ -161,3 +122,185 @@ cluster:
...
+routes:
+  - gateway: {{ include "talm.discovered.default_gateway" . }}
+{{- end }}
+{{- if and .Values.floatingIP (eq .MachineType "controlplane") }}
+---
+apiVersion: v1alpha1
+kind: Layer2VIPConfig
+name: {{ .Values.floatingIP | quote }}
+link: {{ $interfaceName }}
+{{- end }}
```
[the remark this answers is not in the document]
Comment (lexfrei) [reply charts/cozystack/templates/_helpers.tpl:234]: Fixed in cac8dd95. When the default link is a VLAN, `$interfaceName` is rewritten to the parent link only for BondConfig emission; `Layer2VIPConfig.link` now uses a new `$vipLinkName` that points at `$defaultLinkName` (the VLAN), matching the legacy path where the VIP is nested inside the `vlans:` entry.

File: charts/generic/templates/_helpers.tpl [hunk 12]
```diff
@@ -12,6 +21,140 @@ machine:
...
+{{- end }}
+{{- if and .Values.floatingIP (eq .MachineType "controlplane") }}
+---
+apiVersion: v1alpha1
+kind: Layer2VIPConfig
+name: {{ .Values.floatingIP | quote }}
+link: {{ $interfaceName }}
+{{- end }}
```
[the remark this answers is not in the document]
Comment (lexfrei) [reply charts/generic/templates/_helpers.tpl:157]: Fixed in 463bfb77. Same change as the cozystack chart — `Layer2VIPConfig.link` now uses `$defaultLinkName` when the default link is a VLAN, so the VIP stays attached to the configured VLAN.

**Review by myasnikovdaniil — APPROVED** [review]
Re-reviewed. All items from the first review are addressed:

- Malformed TalosVersion crash → fixed (early validation in `Render()`)
- BondConfig/VLANConfig test coverage → fixed (`bondTopologyLookup` + `vlanOnBondTopologyLookup`)
- Layer2VIPConfig positive assertion → fixed
- README note precision → fixed
- machine.common comment → fixed
- Layer2VIP attached to wrong link in VLAN path → fixed (`$vipLinkName`)

Tests pass clean, including with `-race`. Schema matches actual Talos Go types. LGTM.
